From: Gerald Schaefer <geraldsc@de.ibm.com>
Subject: dasd: fix  message flood for unsolicited interrupts
References: bnc#440610

Symptom:     The message log is flooded with messages for unsolicited
             interrupts and missing sense data.
Problem:     CIO generates fake IRBs which are falsly interpreted as
             unsolicited interrupts.
Solution:    Ignore fake IRBs in unsolicited interupt handler.

Acked-by: John Jolly <jjolly@suse.de>
---
 drivers/s390/cio/qdio.h      |    8 ++++++++
 drivers/s390/cio/qdio_main.c |    6 ++++++
 2 files changed, 14 insertions(+)

Index: linux-sles11/drivers/s390/cio/qdio.h
===================================================================
--- linux-sles11.orig/drivers/s390/cio/qdio.h
+++ linux-sles11/drivers/s390/cio/qdio.h
@@ -16,6 +16,14 @@
 #define QDIO_BUSY_BIT_GIVE_UP		2000000	/* 2 seconds = eternity */
 #define QDIO_INPUT_THRESHOLD		500	/* 500 microseconds */
 
+/*
+ * if an asynchronous HiperSockets queue runs full, the 10 seconds timer wait
+ * till next initiative to give transmitted skbs back to the stack is too long.
+ * Therefore polling is started in case of multicast queue is filled more
+ * than 50 percent.
+ */
+#define QDIO_IQDIO_POLL_LVL		65	/* HS multicast queue */
+
 enum qdio_irq_states {
 	QDIO_IRQ_STATE_INACTIVE,
 	QDIO_IRQ_STATE_ESTABLISHED,
Index: linux-sles11/drivers/s390/cio/qdio_main.c
===================================================================
--- linux-sles11.orig/drivers/s390/cio/qdio_main.c
+++ linux-sles11/drivers/s390/cio/qdio_main.c
@@ -851,6 +851,12 @@ static void __qdio_outbound_processing(s
 	if (queue_type(q) == QDIO_IQDIO_QFMT && !multicast_outbound(q))
 		return;
 
+	if ((queue_type(q) == QDIO_IQDIO_QFMT) &&
+	    (atomic_read(&q->nr_buf_used)) > QDIO_IQDIO_POLL_LVL) {
+		tasklet_schedule(&q->tasklet);
+		return;
+	}
+
 	if (q->u.out.pci_out_enabled)
 		return;
 
